Instructor Email Distribution Tool Implementation

caijim opened the pull request:
Problem: Instructors need a way to
e-mail students based on specific criteria.

Solution: Front-end table to select students,
back-end bulk data model querying. Back-end unit tests done,
front-end unit tests to come

Opening this a little early (code is finalized and we have been running it in the stanford fork for months, but front end tests need to be added). My ra-ship is completed around the first week of June so it would be nice if we could get this into master by then (or at least be in good shape). Any and all comments welcome. Thank you!

![emailwidget](https://cloud.githubusercontent.com/assets/461386/7600143/6e4caada-f8bd-11e4-96db-2ad5ad33d7e9.gif)

edx-webhook commented:
Thanks for the pull request, @caijim! I've created [OSPR-586](https://openedx.atlassian.net/browse/OSPR-586) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

We can't start reviewing your pull request until you've added yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS) file. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

sarina commented:
@caijim - do you have any documentation on this feature? It would be helpful to have either a document of detailed test steps (so we can spin up the feature locally and walk through it) Screenshots at each step would be a bonus, so we can see what UX you're expecting to present.

caijim commented:
Sorry, I accidentally closed this ticket (and I presume the auto-jira admin closed it). Can you re-assign the status please? I'll work on the documentation

sarina commented:
@caijim np, done. Be sure you're up to date with the most current version of edx-platform master - we've had some test issues and changes that were merged yesterday & today.

caijim commented:
I added documentation https://docs.google.com/document/d/1RIcwPmUjIEUwP1ymwGhRh3_3bEcUP20b4IvNt4pvtag/edit

And rebased on the latest version of edx/master

sarina commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -2448,3 +2453,443 @@ def mark_student_can_skip_entrance_exam(request, course_id):  # pylint: disable=
         'message': message,
     }
     return JsonResponse(response_payload)
+
```
This file is getting super crazy long. I wonder if it's worth putting some of these methods into a helper file of some sort.

sarina commented:
As mentioned in email to Greg Bruns, I'm going to close this PR until we can come to a consensus on how we want this feature and review process to work, and we're all ready to jump in on reviewing this. If you have questions please reach out to me. Thanks.
